vimPlugins: backoff on timeout in update.py

timokau opened the pull request:
###### Motivation for this change

Updating vim-plugins recently started timing out regularly for me. It
may have to do with an ISP switch on my side, but I don't think that
should cause timeouts. I guess I'm probably not the only one
experiencing this, so in this comment I introduce exponential backoff.
Every request will be retried up to 3 times (3 seconds, 6 seconds, 12
seconds delay).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

CC @Mic92 @rvolosatovs

Mic92 commented on pkgs/misc/vim-plugins/update.py:
```diff
@@ -24,12 +24,59 @@
+from functools import wraps
+
+
+def retry(ExceptionToCheck, tries=4, delay=3, backoff=2, logger=None):
```
Can you add type hints here as well?

Mic92 replied:
```suggestion
def retry(ExceptionToCheck: Exception, tries: int = 4, delay : int = 3, backoff: int = 2):
```

Mic92 replied:
All the type hints are untested, so you may want to check with mypy.

timokau replied:
Unfortunately `Exception` didn't work, maybe python special cases that? mypy complained:

```
update.py:58: error: Exception type must be derived from BaseException
update.py:81: error: Argument 1 to "retry" has incompatible type "Type[URLError]"; expected "Exception"
update.py:94: error: Argument 1 to "retry" has incompatible type "Type[URLError]"; expected "Exception"
```

So I used "Any" instead (which also respects that you could theoretically pass in a tuple of Exceptions).

Mic92 commented on pkgs/misc/vim-plugins/update.py:
```diff
@@ -24,12 +24,59 @@
+    :param logger: logger to use. If None, print
+    :type logger: logging.Logger instance
+    """
+    def deco_retry(f):
```
And here.

Mic92 replied:
```suggestion
    def deco_retry(f: Callable) -> Callable:
```

Mic92 commented on pkgs/misc/vim-plugins/update.py:
```diff
@@ -24,12 +24,59 @@
+    def deco_retry(f):
+
+        @wraps(f)
+        def f_retry(*args, **kwargs):
```
```suggestion
        def f_retry(*args: Any, **kwargs: Any) -> Any:
```

Mic92 commented on pkgs/misc/vim-plugins/update.py:
```diff
@@ -24,12 +24,59 @@
+    :param backoff: backoff multiplier e.g. value of 2 will double the delay
+        each retry
+    :type backoff: int
+    :param logger: logger to use. If None, print
```
The logger is unused and can be removed.

timokau committed:
vimPlugins: fix formatting and typing in update.py

timokau committed:
vimPlugins: backoff on timeout in update.py

Updating vim-plugins recently started timing out regularly for me. It
may have to do with an ISP switch on my side, but I don't think that
should cause timeouts. I guess I'm probably not the only one
experiencing this, so in this comment I introduce exponential backoff.
Every request will be retried up to 3 times (3 seconds, 6 seconds, 12
seconds delay).

timokau commented:
Addressed your comments. Also fixed unrelated formatting and typing issues.

Mic92 commented:
Unfortunately we now have a merging conflict.

timokau commented:
Rebased.

timokau commented:
@Mic92 please have another look :)
